src/plugins/discover/public/application/components/discover_grid/discover_grid.tsx

```diff
@@ -0,0 +1,329 @@
+
+export const EuiDataGridMemoized = React.memo((props: EuiDataGridProps) => {
+  return <EuiDataGrid {...props} />;
+});
+
+export const DiscoverGrid = ({
+  ariaLabelledBy,
+  columns,
+  defaultColumns,
+  indexPattern,
+  onAddColumn,
+  onFilter,
+  onRemoveColumn,
+  onResize,
+  onSetColumns,
+  onSort,
+  rows,
+  sampleSize,
+  searchDescription,
+  searchTitle,
+  services,
+  settings,
+  showTimeCol,
+  sort,
+}: DiscoverGridProps) => {
+  const [expanded, setExpanded] = useState<ElasticSearchHit | undefined>(undefined);
+
+  /**
+   * Pagination
+   */
+  const [pagination, setPagination] = useState({ pageIndex: 0, pageSize: defaultPageSize });
+
+  const rowsLength = rows ? rows.length : 0;
+  const pageCount = Math.ceil(rowsLength / pagination.pageSize);
+  const isOnLastPage = pagination.pageIndex === pageCount - 1;
+
+  const paginationObj = useMemo(() => {
+    const onChangeItemsPerPage = (pageSize: number) =>
+      setPagination((paginationData) => ({ ...paginationData, pageSize }));
+
+    const onChangePage = (pageIndex: number) =>
+      setPagination((paginationData) => ({ ...paginationData, pageIndex }));
+
+    return {
+      onChangeItemsPerPage,
+      onChangePage,
+      pageIndex: pagination.pageIndex > pageCount - 1 ? 0 : pagination.pageIndex,
+      pageSize: pagination.pageSize,
+      pageSizeOptions: pageSizeArr,
+    };
+  }, [pagination, pageCount]);
+
+  /**
+   * Sorting
+   */
+  const sortingColumns = useMemo(() => sort.map(([id, direction]) => ({ id, direction })), [sort]);
+
+  const onTableSort = useCallback(
+    (sortingColumnsData) => {
+      onSort(sortingColumnsData.map(({ id, direction }: SortObj) => [id, direction]));
+    },
+    [onSort]
+  );
+
+  /**
+   * Cell rendering
+   */
+  const renderCellValue = useMemo(() => getRenderCellValueFn(indexPattern, rows), [
+    rows,
+    indexPattern,
+  ]);
+
+  /**
+   * Render variables
+   */
+  const showDisclaimer = rowsLength === sampleSize && isOnLastPage;
+  const randomId = useMemo(() => htmlIdGenerator()(), []);
+
+  const rowCount = useMemo(() => (rows ? rows.length : 0), [rows]);
+  const euiGridColumns = useMemo(
+    () => getEuiGridColumns(columns, settings, indexPattern, showTimeCol, defaultColumns),
+    [columns, indexPattern, showTimeCol, settings, defaultColumns]
+  );
+  const schemaDetectors = useMemo(() => getSchemaDetectors(), []);
+  const popoverContents = useMemo(() => getPopoverContents(), []);
+  const columnsVisibility = useMemo(
+    () => ({
+      visibleColumns: getVisibleColumns(columns, indexPattern, showTimeCol) as string[],
+      setVisibleColumns: (newColumns: string[]) => {
+        onSetColumns(newColumns);
+      },
+    }),
+    [columns, indexPattern, showTimeCol, onSetColumns]
+  );
+  const sorting = useMemo(() => ({ columns: sortingColumns, onSort: onTableSort }), [
+    sortingColumns,
+    onTableSort,
+  ]);
+  const lead = useMemo(() => getLeadControlColumns(), []);
+
+  if (!rowCount || !rows) {
```

Comment:
As rowCount has the check for `rows` I think that the second check is not needed but maybe I am missing something here

Reply:
You're right, guess it was here to satisfy the type check of the context that's created later in the file, but this can be done differently, so, removed it